File: game/garden.py
"""
Garden Class
Manages the grid-based garden system and plant placement
"""

from typing import Dict, List, Optional, Tuple
from constants import *
from plant import Plant

class Garden:
    """Grid-based garden system for growing plants"""
    
    def __init__(self):
        self.width = GRID_WIDTH
        self.height = GRID_HEIGHT
        
        # Grid system
        self.grid = [[None for _ in range(self.width)] for _ in range(self.height)]
        self.plants = {}  # (x, y) -> Plant mapping
        
        # Garden state
        self.soil_quality = [[1.0 for _ in range(self.width)] for _ in range(self.height)]
        self.water_levels = [[0 for _ in range(self.width)] for _ in range(self.height)]
        self.fertilizer_levels = [[0 for _ in range(self.width)] for _ in range(self.height)]
        
        # Garden expansions
        self.expansions = 1
        self.max_expansions = 3
        
        # Weather effects
        self.rain_timer = 0.0
        self.pest_infestation = False
        self.pest_timer = 0.0
        
        # Initialize starting garden area
        self._initialize_garden()
    
    def _initialize_garden(self):
        """Initialize the starting garden area"""
        # Start with a 5x5 garden in the center
        start_x = (self.width // 2) - 2
        start_y = (self.height // 2) - 2
        
        for y in range(start_y, start_y + 5):
            for x in range(start_x, start_x + 5):
                if 0 <= x < self.width and 0 <= y < self.height:
                    self.soil_quality[y][x] = 1.0
                    self.water_levels[y][x] = 2
                    self.fertilizer_levels[y][x] = 0
    
    def update(self, dt: float):
        """Update garden state"""
        # Update all plants
        for plant in self.plants.values():
            plant.update(dt)
        
        # Update weather effects
        if self.rain_timer > 0:
            self.rain_timer -= dt
            if self.rain_timer <= 0:
                self._end_rain_effect()
        
        # Update pest infestation
        if self.pest_infestation:
            self.pest_timer += dt
            if self.pest_timer >= 30.0:  # Pests last 30 seconds
                self._end_pest_infestation()
    
    def plant_seed(self, x: int, y: int, seed_type: str) -> bool:
        """Plant a seed at the specified grid position"""
        if not self._is_valid_position(x, y):
            return False
        
        if self.grid[y][x] is not None:
            return False  # Position already occupied
        
        if not self._is_plantable_soil(x, y):
            return False  # Not plantable soil
        
        # Create new plant
        plant = Plant(seed_type, x, y)
        self.grid[y][x] = plant
        self.plants[(x, y)] = plant
        
        # Apply soil quality effects
        soil_quality = self.soil_quality[y][x]
        if soil_quality < 0.5:
            plant.growth_progress *= 0.8  # Slower growth on poor soil
        
        return True
    
    def water_plant(self, x: int, y: int) -> bool:
        """Water a plant at the specified position"""
        if not self._is_valid_position(x, y):
            return False
        
        plant = self.grid[y][x]
        if plant is None:
            return False
        
        # Water the plant
        if plant.water():
            # Update soil water level
            self.water_levels[y][x] = min(5, self.water_levels[y][x] + 1)
            return True
        
        return False
    
    def fertilize_plant(self, x: int, y: int) -> bool:
        """Apply fertilizer to a plant"""
        if not self._is_valid_position(x, y):
            return False
        
        plant = self.grid[y][x]
        if plant is None:
            return False
        
        # Apply fertilizer
        if plant.fertilize():
            # Update soil fertilizer level
            self.fertilizer_levels[y][x] = min(3, self.fertilizer_levels[y][x] + 1)
            return True
        
        return False
    
    def get_plant(self, x: int, y: int) -> Optional[Plant]:
        """Get plant at specified position"""
        if not self._is_valid_position(x, y):
            return None
        return self.grid[y][x]
    
    def remove_plant(self, x: int, y: int) -> bool:
        """Remove plant from specified position"""
        if not self._is_valid_position(x, y):
            return False
        
        if self.grid[y][x] is not None:
            plant = self.grid[y][x]
            del self.plants[(x, y)]
            self.grid[y][x] = None
            
            # Improve soil quality slightly when plant is harvested
            self.soil_quality[y][x] = min(1.5, self.soil_quality[y][x] + 0.1)
            
            return True
        
        return False
    
    def expand(self) -> bool:
        """Expand the garden area"""
        if self.expansions >= self.max_expansions:
            return False
        
        self.expansions += 1
        
        # Increase plantable area
        expansion_size = 2
        center_x = self.width // 2
        center_y = self.height // 2
        
        for y in range(center_y - expansion_size, center_y + expansion_size + 1):
            for x in range(center_x - expansion_size, center_x + expansion_size + 1):
                if 0 <= x < self.width and 0 <= y < self.height:
                    if self.soil_quality[y][x] == 0:
                        self.soil_quality[y][x] = 0.8
                        self.water_levels[y][x] = 1
        
        return True
    
    def apply_rain_effect(self):
        """Apply rain effect to all plants"""
        self.rain_timer = 20.0  # Rain lasts 20 seconds
        
        # Water all plants
        for plant in self.plants.values():
            plant.water()
        
        # Increase soil water levels
        for y in range(self.height):
            for x in range(self.width):
                if self._is_plantable_soil(x, y):
                    self.water_levels[y][x] = min(5, self.water_levels[y][x] + 2)
    
    def _end_rain_effect(self):
        """End rain effect"""
        self.rain_timer = 0.0
    
    def trigger_pest_infestation(self):
        """Trigger pest infestation event"""
        if not self.pest_infestation:
            self.pest_infestation = True
            self.pest_timer = 0.0
            
            # Randomly destroy some plants
            import random
            plant_positions = list(self.plants.keys())
            if plant_positions:
                num_to_destroy = min(3, len(plant_positions))
                positions_to_destroy = random.sample(plant_positions, num_to_destroy)
                
                for x, y in positions_to_destroy:
                    self.remove_plant(x, y)
    
    def _end_pest_infestation(self):
        """End pest infestation"""
        self.pest_infestation = False
        self.pest_timer = 0.0
    
    def _is_valid_position(self, x: int, y: int) -> bool:
        """Check if position is within garden bounds"""
        return 0 <= x < self.width and 0 <= y < self.height
    
    def _is_plantable_soil(self, x: int, y: int) -> bool:
        """Check if position has plantable soil"""
        return self.soil_quality[y][x] > 0
    
    def get_garden_summary(self) -> Dict:
        """Get summary of garden state"""
        return {
            "width": self.width,
            "height": self.height,
            "expansions": self.expansions,
            "max_expansions": self.max_expansions,
            "total_plants": len(self.plants),
            "plantable_tiles": sum(1 for row in self.soil_quality for quality in row if quality > 0),
            "weather": {
                "rain_timer": self.rain_timer,
                "pest_infestation": self.pest_infestation,
                "pest_timer": self.pest_timer
            }
        }
    
    def get_plant_positions(self) -> List[Tuple[int, int]]:
        """Get list of all plant positions"""
        return list(self.plants.keys())
    
    def get_plantable_positions(self) -> List[Tuple[int, int]]:
        """Get list of all plantable positions"""
        positions = []
        for y in range(self.height):
            for x in range(self.width):
                if self._is_plantable_soil(x, y):
                    positions.append((x, y))
        return positions

File: web/js/garden.js
// Garden Class - Manages the grid-based garden system and plant placement
class Garden {
    constructor() {
        this.width = GRID_WIDTH;
        this.height = GRID_HEIGHT;
        
        // Grid system
        this.grid = Array(this.height).fill().map(() => Array(this.width).fill(null));
        this.plants = new Map(); // (x, y) -> Plant mapping
        
        // Garden state
        this.soilQuality = Array(this.height).fill().map(() => Array(this.width).fill(1.0));
        this.waterLevels = Array(this.height).fill().map(() => Array(this.width).fill(0));
        this.fertilizerLevels = Array(this.height).fill().map(() => Array(this.width).fill(0));
        
        // Garden expansions
        this.expansions = 1;
        this.maxExpansions = 3;
        
        // Weather effects
        this.rainTimer = 0.0;
        this.pestInfestation = false;
        this.pestTimer = 0.0;
        
        // Initialize starting garden area
        this.initializeGarden();
    }
    
    initializeGarden() {
        // Start with a 5x5 garden in the center
        const startX = Math.floor(this.width / 2) - 2;
        const startY = Math.floor(this.height / 2) - 2;
        
        for (let y = startY; y < startY + 5; y++) {
            for (let x = startX; x < startX + 5; x++) {
                if (x >= 0 && x < this.width && y >= 0 && y < this.height) {
                    this.soilQuality[y][x] = 1.0;
                    this.waterLevels[y][x] = 2;
                    this.fertilizerLevels[y][x] = 0;
                }
            }
        }
    }
    
    update(dt) {
        // Update all plants
        for (const plant of this.plants.values()) {
            plant.update(dt);
        }
        
        // Update weather effects
        if (this.rainTimer > 0) {
            this.rainTimer -= dt;
            if (this.rainTimer <= 0) {
                this.endRainEffect();
            }
        }
        
        // Update pest infestation
        if (this.pestInfestation) {
            this.pestTimer += dt;
            if (this.pestTimer >= 30.0) { // Pests last 30 seconds
                this.endPestInfestation();
            }
        }
    }
    
    plantSeed(x, y, seedType) {
        if (!this.isValidPosition(x, y)) {
            return false;
        }
        
        if (this.grid[y][x] !== null) {
            return false; // Position already occupied
        }
        
        if (!this.isPlantableSoil(x, y)) {
            return false; // Not plantable soil
        }
        
        // Create new plant
        const plant = new Plant(seedType, x, y);
        this.grid[y][x] = plant;
        this.plants.set(`${x},${y}`, plant);
        
        // Apply soil quality effects
        const soilQuality = this.soilQuality[y][x];
        if (soilQuality < 0.5) {
            plant.growthProgress *= 0.8; // Slower growth on poor soil
        }
        
        return true;
    }
    
    waterPlant(x, y) {
        if (!this.isValidPosition(x, y)) {
            return false;
        }
        
        const plant = this.grid[y][x];
        if (plant === null) {
            return false;
        }
        
        // Water the plant
        if (plant.water()) {
            // Update soil water level
            this.waterLevels[y][x] = Math.min(5, this.waterLevels[y][x] + 1);
            return true;
        }
        
        return false;
    }
    
    fertilizePlant(x, y) {
        if (!this.isValidPosition(x, y)) {
            return false;
        }
        
        const plant = this.grid[y][x];
        if (plant === null) {
            return false;
        }
        
        // Apply fertilizer
        if (plant.fertilize()) {
            // Update soil fertilizer level
            this.fertilizerLevels[y][x] = Math.min(3, this.fertilizerLevels[y][x] + 1);
            return true;
        }
        
        return false;
    }
    
    getPlant(x, y) {
        if (!this.isValidPosition(x, y)) {
            return null;
        }
        return this.grid[y][x];
    }
    
    removePlant(x, y) {
        if (!this.isValidPosition(x, y)) {
            return false;
        }
        
        if (this.grid[y][x] !== null) {
            const plant = this.grid[y][x];
            this.plants.delete(`${x},${y}`);
            this.grid[y][x] = null;
            
            // Improve soil quality slightly when plant is harvested
            this.soilQuality[y][x] = Math.min(1.5, this.soilQuality[y][x] + 0.1);
            
            return true;
        }
        
        return false;
    }
    
    expand() {
        if (this.expansions >= this.maxExpansions) {
            return false;
        }
        
        this.expansions++;
        
        // Increase plantable area
        const expansionSize = 2;
        const centerX = Math.floor(this.width / 2);
        const centerY = Math.floor(this.height / 2);
        
        for (let y = centerY - expansionSize; y <= centerY + expansionSize; y++) {
            for (let x = centerX - expansionSize; x <= centerX + expansionSize; x++) {
                if (x >= 0 && x < this.width && y >= 0 && y < this.height) {
                    if (this.soilQuality[y][x] === 0) {
                        this.soilQuality[y][x] = 0.8;
                        this.waterLevels[y][x] = 1;
                    }
                }
            }
        }
        
        return true;
    }
    
    applyRainEffect() {
        this.rainTimer = 20.0; // Rain lasts 20 seconds
        
        // Water all plants
        for (const plant of this.plants.values()) {
            plant.water();
        }
        
        // Increase soil water levels
        for (let y = 0; y < this.height; y++) {
            for (let x = 0; x < this.width; x++) {
                if (this.isPlantableSoil(x, y)) {
                    this.waterLevels[y][x] = Math.min(5, this.waterLevels[y][x] + 2);
                }
            }
        }
    }
    
    endRainEffect() {
        this.rainTimer = 0.0;
    }
    
    triggerPestInfestation() {
        if (!this.pestInfestation) {
            this.pestInfestation = true;
            this.pestTimer = 0.0;
            
            // Randomly destroy some plants
            const plantPositions = Array.from(this.plants.keys());
            if (plantPositions.length > 0) {
                const numToDestroy = Math.min(3, plantPositions.length);
                const positionsToDestroy = this.shuffleArray(plantPositions).slice(0, numToDestroy);
                
                for (const pos of positionsToDestroy) {
                    const [x, y] = pos.split(',').map(Number);
                    this.removePlant(x, y);
                }
            }
        }
    }
    
    endPestInfestation() {
        this.pestInfestation = false;
        this.pestTimer = 0.0;
    }
    
    isValidPosition(x, y) {
        return x >= 0 && x < this.width && y >= 0 && y < this.height;
    }
    
    isPlantableSoil(x, y) {
        return this.soilQuality[y][x] > 0;
    }
    
    shuffleArray(array) {
        const shuffled = [...array];
        for (let i = shuffled.length - 1; i > 0; i--) {
            const j = Math.floor(Math.random() * (i + 1));
            [shuffled[i], shuffled[j]] = [shuffled[j], shuffled[i]];
        }
        return shuffled;
    }
    
    getGardenSummary() {
        return {
            width: this.width,
            height: this.height,
            expansions: this.expansions,
            maxExpansions: this.maxExpansions,
            totalPlants: this.plants.size,
            plantableTiles: this.soilQuality.flat().filter(quality => quality > 0).length,
            weather: {
                rainTimer: this.rainTimer,
                pestInfestation: this.pestInfestation,
                pestTimer: this.pestTimer
            }
        };
    }
    
    getPlantPositions() {
        return Array.from(this.plants.keys()).map(pos => {
            const [x, y] = pos.split(',').map(Number);
            return [x, y];
        });
    }
    
    getPlantablePositions() {
        const positions = [];
        for (let y = 0; y < this.height; y++) {
            for (let x = 0; x < this.width; x++) {
                if (this.isPlantableSoil(x, y)) {
                    positions.push([x, y]);
                }
            }
        }
        return positions;
    }
}
